staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -228,7 +228,31 @@ func (dm *discoveryManager) fetchFreshDiscoveryForService(gv metav1.GroupVersion
 
 		dm.setCacheEntryForService(info.service, cached)
 		return &cached, nil
-	case http.StatusNotAcceptable:
+	} else if writer.respCode == http.StatusOK && isV2Beta1ContentType(writer.Header().Get("Content-Type")) {
+		parsed := &apidiscoveryv2beta1.APIGroupDiscoveryList{}
+		if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), writer.data, parsed); err != nil {
+			return nil, err
+		}
+
+		klog.V(3).Infof("DiscoveryManager: Successfully downloaded discovery for %s", info.service.String())
+
+		// Convert discovery info into a map for convenient lookup later
+		discoMap := map[metav1.GroupVersion]apidiscoveryv2beta1.APIVersionDiscovery{}
+		for _, g := range parsed.Items {
+			for _, v := range g.Versions {
+				discoMap[metav1.GroupVersion{Group: g.Name, Version: v.Version}] = v
+			}
+		}
+
+		// Save cached result
+		cached = cachedResult{
+			discovery:   discoMap,
+			etag:        writer.Header().Get("Etag"),
+			lastUpdated: now,
+		}
+		dm.setCacheEntryForService(info.service, cached)
+		return &cached, nil
+	} else {
 		// Discovery Document is not being served at all.
```

Comment:
One notable side effect of using legacy as a catch-all fallback for any error is that if the server is just completely non-responsive, or times out, or doesnt speak HTTP we will then try legacy which is bound to also fail.

Has this been considered?

Reply:
normally a request level error (dial, protocol, timeout errors, etc) is distinguishable from a status code error response, but it doesn't look like the ServeHTTP path writing to an in-memory response writer preserves that distinction… that's really unfortunate

Reply:
it looks like in timeout cases, the TimeoutWriter will write a StatusServiceUnavailable code. handling that case first and not falling back to legacy discovery would make sense to me